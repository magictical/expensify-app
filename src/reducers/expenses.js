// Expenses Reducer
const expenseReducerDefaultState = [];
// Reducer for expenses
export default (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                //spread operator 사용 ...은 이전의 array를 불러오는것
                // 즉 이전 state array, action.expense를 이어서 붙이는것(삽입) 
                ...state, 
                action.expense
            ]
        //filter의 파라메터 id는 {}를 사용해서 desructuring 됨 즉 state의 천번째 
        //property인 id를 직적 지정한거임 이걸 action.id와 비교해서 참이면 통과
        //거짓이면 필터링 즉 state필터의 파라메터와 입력된 id값이 동일하면 filter의 
        //새로운 array에서 제외된다.
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id )
        case 'EDIT_EXPENSE' :
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        //read all the expense data from state
                        ...state,
                        //override updates data in expense
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

