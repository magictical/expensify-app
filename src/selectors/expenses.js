// filter for expense
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        //리턴값에 항상 true를 전달하기 위해 어떤값을 입력받든 true가되는 식을만듬
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        
        const textMatch =  expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;

    //check sort function in MDN https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
    }).sort((a, b) => {
        // 비교후 -1이면 a의 인덱스를 작은값으로 즉 a가 앞으로정렬 1이면 a인덱스를 큰값으로
        if(sortBy === 'date') {
            //최신 순으로 정렬
            return a.createdAt < b.createdAt ? 1 : -1
        }
        if(sortBy === 'amount') {
            // amount가 높은 순으로 정렬
            return a.amount > b.amount ? -1: 1
        }
    });
}
