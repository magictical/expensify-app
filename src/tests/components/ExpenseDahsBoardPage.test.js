import React from 'react';
import ExpenseDashboardPage from '../../components/ExpenseDashBoardPage';
// import ExpenseList from '../../components/ExpenseList';
// import ExpenseListFilter from '../../components/ExpenseListFilters'
import { shallow } from 'enzyme';
import expenses from '../fixture/expenses';

test('should render dashboard with filters and list', () => {
  const wrapper = shallow(<ExpenseDashboardPage />);
  expect(wrapper).toMatchSnapshot();
});
