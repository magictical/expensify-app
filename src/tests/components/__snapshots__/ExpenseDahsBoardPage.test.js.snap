// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render dashboard with filters and list 1`] = `
<div>
  <p>
    This is from my dashboard component
  </p>
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;
